src/RequestQueue.js

```diff
@@ -1,34 +1,83 @@
-'use strict'; 
+'use strict';
 
 import SuperAgent from 'superagent';
 
 function RequestQueue( superAgent ) {
-	
+
 	let Request = superAgent.Request;
 	let superAgentEnd = Request.prototype.end;
-
 	let queue = [];
 
-	//Reset request to allow retry
-	function reset( obj, timeout ) {
-		
-		delete obj.xhr;
-		delete obj._timer;
-		delete obj.aborted;
-		delete obj.timedout;
-		delete obj.timeout;
+	const TIMEOUT_REGEX = /timeout of \d+ms exceeded/;
+	const CORS_REGEX = /Origin is not allowed by Access-Control-Allow-Origin/;
+
+	let RETRY_TIMEOUT;
+
+	//HACK to Reset request to allow retry
+	function _resetRequest( request, timeout ) {
+
+		let headers = {};
+
+		if ( request.req ) {
+			headers = request.req ? request.req._headers : {};
+			request.req.abort();
+		}
+
+		request.called = false;
+		request.timeout( timeout );
+
+		delete request._timer;
+		delete request._aborted;
+		delete request.timedout;
+		delete request.req;
+		delete request.xhr;
+
+		let headerKeys = Object.keys( headers );
 
-		obj.timeout( timeout );
+		for( let i = 0; i < headerKeys.length; i++ ) {
+			request.set( headerKeys[i], headers[headerKeys[i]] );
+		}
 	}
 
-	function handleConnError( connErrorHandler, err ) {
-		
-		if( connErrorHandler ) {
-			connErrorHandler( err );
+	function _returnResponse( fn, err, res ) {
+		if ( fn ) {
+			fn( err, res );
 		}
 	}
 
-	function pop() {
+	function _handleConnectionError( connectionErrorHandler, err ) {
+		if ( connectionErrorHandler ) {
+			connectionErrorHandler( err );
+		}
+	}
+
+	function _shouldAttempRetry( err, retryEnabled ) {
+
+		if ( !retryEnabled || !err) {
+			return false;
+		}
+
+		let gatewayOrServiceUnavailable =
+				err.status && ( err.status === 502 || err.status === 503 || err.status === 504 );
+
+		let requestTimedOut = TIMEOUT_REGEX.test( err ) || err.code ==='ECONNABORTED';
+
+		if ( gatewayOrServiceUnavailable || requestTimedOut ) {
+			RETRY_TIMEOUT = 2000;
+			return true;
+		}
+
+		let corsError = CORS_REGEX.test( err );

```

Comment:
This CORS check is required due to the fact that if you lose internet connection and try to contact LORES it will fail on the OPTIONS request and throw a generic CORS error. It is not possible( tried a bunch of different things) to decypher what problem behind the CORS error was. Superagent just returns "Origin is not allowed by Access-Control-Allow-Origin".